src/solver.rs

```diff
@@ -8,8 +10,67 @@ pub fn query_smt(query: &String) -> bool {
 }
 
 pub fn query_smt_with_solver(query: &String, solver: SolverConfig) -> bool {
-    let query = format!("{}\n{}", query, "(check-sat)");
+    let (output, stderr) = query_smt_internal(&format!("{query}\n(check-sat)\n"), solver);
 
+    match output.as_str() {
+        "sat\n" => true,
+        "unsat\n" => false,
+        _ => panic!(
+            "Invalid output from smt solver.\nQuery: {query}\nStdout: {output}\nStderr: {stderr}",
+        ),
+    }
+}
+
+pub fn query_smt_with_solver_and_eval(
+    query: &String,
+    exprs_to_eval: &Vec<String>,
+    solver: SolverConfig,
+) -> (bool, Vec<BigUint>) {
+    // TODO we are calling the solver twice because we do not know
+    // if the query is sat.
+    if !query_smt_with_solver(query, solver.clone()) {
+        (false, vec![])
+    } else if exprs_to_eval.is_empty() {
+        (true, vec![])
+    } else {
+        let (output, stderr) = query_smt_internal(
+            &format!(
+                "(set-option :produce-models true)\n{query}\n(check-sat)\n(get-value ({}))\n",
```

Comment:
why not do this right away?

Reply:
Because `get-value` can only be used if the immediately preceding `check-sat` returns `sat`

Reply:
sure but you can add anyway and the solver can return
```
unsat
(error not sat)
```
if unsat

Reply:
Yeah...

Reply:
I still think this should be done in a single query